### DISCUSSION TITLE
XState state.exit.actions & micro transitions will be executed in the next state?

### DISCUSSION THREAD

**Discussion by huan** [statelyai/xstate#3231] [General]
For example:

```ts
import { createMachine, interpret } from "xstate";

const m = createMachine({
  initial: "busy",
  states: {
    idle: {},
    busy: {
      after: {
        10: "idle"
      },
      exit: [(_, __, { state }) => console.log("state.value", state.value)]
    }
  }
});

interpret(m).start();
```

> Updated the code from @Andarist at https://codesandbox.io/s/competent-antonelli-f1kvf8?file=/src/index.js

In the above code, the `console.log()` will be executed within the state `idle`, instead of in the state `busy`.

If the `action` is sensitive to the current state, then a workaround will be needed.

My question is: is this an expected behavior or not?

Link to:

- https://github.com/huan/mailbox/issues/4

**Comment by davidkpiano** [comment]
What is the initial state for the machine? Can you make a repro?

**Reply by huan** [reply]
Yes I can. Will make a reproducible repo for it later.

**Reply by Andarist** [reply]
I'm not sure if you wire things up in `mailbox` in the same way, but I've prepared a simple repro case that demonstrates the issue.

https://codesandbox.io/s/competent-antonelli-f1kvf8?file=/src/index.js

This is part of the reason why I think that `meta.state` should be removed/repurposed/reimplemented. At the moment it's confusing and very error-prone. The question is - do you actually want access to the whole `state` object from before the transition or do you just want information on which state node this action was defined?

**Reply by huan** [reply]
@Andarist thank you very much for your help of creating the reproducible code!

My use case of the Mailbox is that it must make sure the actions inside the `states.busy.exit` are executed inside the `states.busy` so that all actors' states are well managed and correct.

From my understanding from an XState user perspective, I think that the `states.busy.exit` actions should run inside the current state instead of the next one.

To be explicit, I expected the `exit` should be `preExit`, but actually it is `postExit`.

I feel it would be better if we can change the behavior of the `exit` from `postExit` to `preExit` because it will be felt more natural.

**Reply by huan** [reply]
@Andarist I remembered one of my use cases: https://github.com/huan/mailbox/issues/4#issuecomment-1107892417

**Reply by huan** [reply]
I found another race condition problem related to this discussion:

- https://github.com/wechaty/bot5-assistant/issues/11
